### PR TITLE
Fix screen-space distortion of bumpwater shader where it samples from…

### PR DESCRIPTION
**Pull request by Beherith**
… stuff closer in Z

The edge of things in front of water also gets refraction distorted. 

We now check wether the depth buffer of the fragment being sampled in a distorted way is actually in front of the water, and if it is, we replace the distortion sample with the original, undistorted screencopy.

### CODE REVIEW

**Comment by Beherith** [comment]
Before:
screen_2024-02-26_12-41-09-692

**Comment by Beherith** [comment]
its just one of the things about bumpwater that have annoyed me for ever.

**Comment by lhog** [comment]
Comparing the screenshots one by one I notice that the intensity of leaves changes on the trees that are unrelated to the refraction pass. Do you know why is that?

**Comment by Beherith** [comment]
LOS shader being different after a reinitialization. Trees are shaded with infotex,

**Comment by lhog** [comment]
Ok, the net effect of the change is quite tiny then.

**Comment by Beherith** [comment]
> Ok, the net effect of the change is quite tiny then.

Yes its tiny, but its a very annoying visual artifact on all edges that are in front of water.

**Comment by lhog** [comment]
Yep, agree. Have a look at the other comment I mentioned in the code and lets merge it then.

**Comment by Beherith** [comment]
I dont see the other comment you left in the code :(

**Comment by lhog** [comment]
Should be visible here: https://github.com/beyond-all-reason/spring/pull/1327/files/e800dfa82009a18e1bda3e3e206571060ef18077

**Comment by sprunk** [comment]
I don't see anything either

File: cont/base/springcontent/shaders/GLSL/BumpWaterFS.glsl [hunk 1]
```diff
@@ -295,8 +295,25 @@ void main()
...
+
+    // Sample the screencopy for refraction at the distorted UV coordinate:
+    vec3 refrColor = texture2D(refraction, refrUV).rgb;
+
+    // Look up the depth of the distorted UV coordinate
+    float  texZ = ConvertDepthToEyeZ(texture2D(depthmap, refrUV).r);
+    float fragZ = ConvertDepthToEyeZ(gl_FragCoord.z);
+
+   // If the distorted coordinate is closer to the screen in Z, then set the mixback to 1.0
+    float mixback = clamp(texZ - fragZ, 0.0, 1.0);
```
Comment (lhog) [inline cont/base/springcontent/shaders/GLSL/BumpWaterFS.glsl:308]: Shouldn't this be some sort of step or smoothstep function?
This one works with one elmo camera unit as a rate for the mix which looks a bit odd.

**Comment by lhog** [comment]
Ok, added the comment in a different way.

File: cont/base/springcontent/shaders/GLSL/BumpWaterFS.glsl (hunk 1 above)
Comment (Beherith) [reply cont/base/springcontent/shaders/GLSL/BumpWaterFS.glsl:308]: Meh, its good enough imo

Comment (Beherith) [reply cont/base/springcontent/shaders/GLSL/BumpWaterFS.glsl:308]: Hardstep :p

**Pull request merged**